### PR TITLE
Remove reagent ids and use typepaths where applicable

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
This replaces /datum/reagent id's with the typepath itself, and removes the superfluous id

TODO/TEST:
- [X] Chem dispensers 
- [X] Some support for legacy macros
- [X] Reactions 
- [x] Chem masters 
- [x] Virus cures
- [x] Grinding
- [x] Drink icons

## Changelog
:cl: Naksu
code: reagent IDs have been removed in favor using reagent typepaths where applicable
fix: mechas, borg hyposprays etc no longer display internal reagent ids to the player
/:cl:

### CODE REVIEW

**Comment by yoyobatty** [comment]
But for what reason?

**Comment by vuonojenmustaturska** [comment]
They're completely unneeded and do not enable any features, they involve frequent lookups to the reagent id = reagent dictionary, they're not compile-time or really any-time checked so typos can and will happen and they're caught when someone makes a bug report about something not working for a month. This has been pretty high up on the to-do list for a while, and i want to get it out of the way before i re-do parts of chemistry.

**Comment by Razharas** [comment]
10/10 10/10
100/100
best pr
BEST PR

**Comment by MrPerson** [comment]
Its rough, but once finished, this has a great chance of being Pull of the Year

File: _maps/RandomRuins/SpaceRuins/TheDerelict.dmm [hunk 1]
```diff
@@ -3372,7 +3372,7 @@
 "lq" = (
 /obj/structure/table,
 /obj/item/reagent_containers/glass/beaker{
-	list_reagents = list("sacid" = 50)
+	list_reagents = list(/datum/reagent/toxin/acid = 50)
```
Comment (ShizCalev) [inline _maps/RandomRuins/SpaceRuins/TheDerelict.dmm]: iirc, the maploader doesn't currently support generating instances from lists (which is why #38300 pops up every once and a while.) 

similar issue also caused #35518

Comment (vuonojenmustaturska) [reply _maps/RandomRuins/SpaceRuins/TheDerelict.dmm]: these are not instances, they are typepaths

**Comment by ninjanomnom** [comment]
Really not a fan of making this change since it removes the support for multiple chem types from the same typepath. Generated chems at runtime to, say, have a different juice for every vegetable and fruit.

**Comment by vuonojenmustaturska** [comment]
condiment bottles were invisible due to using a goofy system for icon_state replacements, i've made it use typelists and fixed the bug

File: code/game/machinery/limbgrower.dm [hunk 2]
```diff
@@ -201,7 +201,7 @@
 
 	for(var/datum/reagent/R in reagents.reagent_list)
 		dat += "[R.name]: [R.volume]"
-		dat += "<A href='?src=[REF(src)];disposeI=[R.id]'>Purge</A><BR>"
```
Comment (AnturK) [inline code/game/machinery/limbgrower.dm]: REF(R) or name ?

File: code/game/objects/effects/decals/cleanable/misc.dm [hunk 3]
```diff
@@ -181,7 +181,7 @@
 
 /obj/effect/decal/cleanable/glitter/white
 	name = "white glitter"
-	icon_state = "nitrous_oxide"
+	icon_state = "no2"
```
Comment (AnturK) [inline code/game/objects/effects/decals/cleanable/misc.dm]: this seems unrelated.

File: code/game/objects/structures/shower.dm [hunk 4]
```diff
@@ -69,7 +69,7 @@
 	cut_overlays()
 
 	if(on)
-		add_overlay(mutable_appearance('icons/obj/watercloset.dmi', "water", ABOVE_MOB_LAYER))
+		add_overlay(mutable_appearance('icons/obj/watercloset.dmi', /datum/reagent/water, ABOVE_MOB_LAYER))
```
Comment (AnturK) [inline code/game/objects/structures/shower.dm]: Doesn't seem correct.

File: code/modules/antagonists/cult/blood_magic.dm [hunk 5]
```diff
@@ -747,7 +747,7 @@
 	var/turf/T = get_turf(target)
 	if(T)
 		for(var/obj/effect/decal/cleanable/blood/B in view(T, 2))
-			if(B.blood_state == "blood")
+			if(B.blood_state == /datum/reagent/blood)
```
Comment (AnturK) [inline code/modules/antagonists/cult/blood_magic.dm]: Is this correct ? Looks like some icon state definition.

File: code/modules/atmospherics/gasmixtures/gas_types.dm [hunk 6]
```diff
@@ -98,7 +98,7 @@ GLOBAL_LIST_INIT(nonreactive_gases, typecacheof(list(/datum/gas/oxygen, /datum/g
 	id = "n2o"
 	specific_heat = 40
 	name = "Nitrous Oxide"
-	gas_overlay = "nitrous_oxide"
+	gas_overlay = /datum/reagent/nitrous_oxide
```
Comment (AnturK) [inline code/modules/atmospherics/gasmixtures/gas_types.dm]: Another one

**Review by AnturK — CHANGES_REQUESTED** [review]
Some find & replace remnants in first half of this.

File: code/modules/food_and_drinks/drinks/drinks.dm [hunk 7]
```diff
@@ -462,28 +462,28 @@
 /obj/item/reagent_containers/food/drinks/soda_cans/cola
 	name = "Space Cola"
 	desc = "Cola. in space."
-	icon_state = "cola"
```
Comment (AnturK) [inline code/modules/food_and_drinks/drinks/drinks.dm]: again

File: code/modules/food_and_drinks/food/condiment.dm [hunk 8]
```diff
@@ -182,15 +187,15 @@
...
+	list_reagents = list(/datum/reagent/consumable/milk = 50)
 	possible_states = list()
 
 /obj/item/reagent_containers/food/condiment/flour
 	name = "flour sack"
 	desc = "A big bag of flour. Good for baking!"
 	icon_state = "flour"
-	item_state = "flour"
-	list_reagents = list("flour" = 30)
+	item_state = /datum/reagent/consumable/flour
```
Comment (AnturK) [inline code/modules/food_and_drinks/food/condiment.dm]: another few in here.

File: code/modules/food_and_drinks/food/condiment.dm [hunk 9]
```diff
@@ -278,16 +293,16 @@
 //Ketchup
 /obj/item/reagent_containers/food/condiment/pack/ketchup
 	name = "ketchup pack"
-	originalname = "ketchup"
-	list_reagents = list("ketchup" = 10)
+	originalname = /datum/reagent/consumable/ketchup
```
Comment (AnturK) [inline code/modules/food_and_drinks/food/condiment.dm]: These seem to be simple text

File: code/modules/food_and_drinks/food/snacks/meat.dm [hunk 10]
```diff
@@ -337,7 +337,7 @@
 
 /obj/item/reagent_containers/food/snacks/meat/steak/gondola
 	name = "gondola steak"
-	tastes = list("meat" = 1, "tranquility" = 1)
+	tastes = list("meat" = 1, /datum/reagent/tranquility = 1)
```
Comment (AnturK) [inline code/modules/food_and_drinks/food/snacks/meat.dm]: These are just descriptions. Also reminds me some test for default named reagents would be nice.

File: code/modules/food_and_drinks/food/snacks_other.dm [hunk 11]
```diff
@@ -83,18 +83,18 @@
 	name = "loaded baked potato"
 	desc = "Totally baked."
 	icon_state = "loadedbakedpotato"
-	bonus_reagents = list("nutriment" = 1, "vitamin" = 2)
-	list_reagents = list("nutriment" = 6)
+	bonus_reagents = list(/datum/reagent/consumable/nutriment = 1, /datum/reagent/consumable/nutriment/vitamin = 2)
+	list_reagents = list(/datum/reagent/consumable/nutriment = 6)
 	filling_color = "#D2B48C"
-	tastes = list("potato" = 1)
+	tastes = list(/datum/reagent/consumable/potato_juice = 1)
```
Comment (AnturK) [inline code/modules/food_and_drinks/food/snacks_other.dm]: More of these, i'll be skipping tastes from now on.

File: code/modules/hydroponics/hydroitemdefines.dm [hunk 12]
```diff
@@ -35,11 +35,11 @@
 	icon = 'icons/obj/hydroponics/equipment.dmi'
 	name = "pest spray"
 	icon_state = "pestspray"
-	item_state = "plantbgone"
+	item_state = /datum/reagent/toxin/plantbgone
```
Comment (AnturK) [inline code/modules/hydroponics/hydroitemdefines.dm]: b

File: code/modules/library/lib_machines.dm [hunk 13]
```diff
@@ -323,14 +323,14 @@ GLOBAL_LIST(cachedbooks) // List of our cached book datums
 	return null
 
 /obj/machinery/computer/libraryconsole/bookmanagement/proc/print_forbidden_lore(mob/user)
-	var/spook = pick("blood", "brass")
+	var/spook = pick(/datum/reagent/blood, "brass")
```
Comment (AnturK) [inline code/modules/library/lib_machines.dm]: this is just someone not knowing about prob

File: code/modules/mining/lavaland/necropolis_chests.dm [hunk 14]
```diff
@@ -882,8 +881,8 @@
 
 /datum/disease/transformation/dragon
 	name = "dragon transformation"
-	cure_text = "nothing"
-	cures = list("adminordrazine")
+	cure_text = /datum/reagent/consumable/nothing
```
Comment (AnturK) [inline code/modules/mining/lavaland/necropolis_chests.dm]: this is just text

File: code/modules/mob/living/silicon/ai/vox_sounds.dm [hunk 15]
```diff
@@ -111,7 +111,7 @@ GLOBAL_LIST_INIT(vox_sounds, list("abduction" = 'sound/vox_fem/abduction.ogg',
 "bleed" = 'sound/vox_fem/bleed.ogg',
 "blob" = 'sound/vox_fem/blob.ogg',
 "blocked" = 'sound/vox_fem/blocked.ogg',
-"blood" = 'sound/vox_fem/blood.ogg',
+/datum/reagent/blood = 'sound/vox_fem/blood.ogg',
```
Comment (AnturK) [inline code/modules/mob/living/silicon/ai/vox_sounds.dm]: c

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 16]
```diff
@@ -299,7 +299,7 @@ Maintenance panel panel is [open ? "opened" : "closed"]"})
 		return 1
 	if(href_list["operation"])
 		switch(href_list["operation"])
-			if("blood")
+			if(/datum/reagent/blood)
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm]: This is preset pda command iirc

File: code/modules/mob/living/simple_animal/slime/slime.dm [hunk 17]
```diff
@@ -73,7 +73,7 @@
 
 	var/static/list/slime_colours = list("rainbow", "grey", "purple", "metal", "orange",
 	"blue", "dark blue", "dark purple", "yellow", "silver", "pink", "red",
-	"gold", "green", "adamantine", "oil", "light pink", "bluespace",
+	"gold", "green", "adamantine", "oil", "light pink", /datum/reagent/bluespace,
```
Comment (AnturK) [inline code/modules/mob/living/simple_animal/slime/slime.dm]: just text

File: code/modules/mob/living/taste.dm [hunk 18]
```diff
@@ -24,7 +24,7 @@
 		if(hallucination > 50 && prob(25))
 			text_output = pick("spiders","dreams","nightmares","the future","the past","victory",\
 			"defeat","pain","bliss","revenge","poison","time","space","death","life","truth","lies","justice","memory",\
-			"regrets","your soul","suffering","music","noise","blood","hunger","the american way")
+			"regrets","your soul","suffering","music","noise",/datum/reagent/blood,"hunger","the american way")
```
Comment (AnturK) [inline code/modules/mob/living/taste.dm]: d

**Review by AnturK — COMMENTED** [review]
More F&R

File: code/modules/antagonists/cult/blood_magic.dm (hunk 5 above)
Comment (vuonojenmustaturska) [reply code/modules/antagonists/cult/blood_magic.dm]: the original code is also wrong, it should be `BLOOD_STATE_HUMAN`

File: code/game/machinery/limbgrower.dm (hunk 2 above)
Comment (vuonojenmustaturska) [reply code/game/machinery/limbgrower.dm]: `[R]` expands to the type in case of datums, there was a missing text2path higher up though

**Comment by vuonojenmustaturska** [comment]
image

**Comment by Altoids1** [comment]
I fucking hate this PR

**Comment by optimumtact** [comment]
long live the king

**Comment by japisd** [comment]
@Altoids1 why?

**Comment by optimumtact** [comment]
downstreams have to rewrite all their reagent code
